app/master/cluster_master.py

```diff
@@ -62,12 +62,34 @@ def api_representation(self):
             'slaves': slaves_representation,
         }
 
-    def builds(self):
+    def builds(self, offset: int=None, limit: int=None) -> List['Build']:
         """
-        Returns a list of all builds
-        :rtype: list[Build]
+        Returns a list of all builds.
+        :param offset: The starting index of the requested build
+        :param limit: The number of builds requested
         """
-        return self._all_builds_by_id.values()
+        num_builds = len(self._all_builds_by_id)
+        highest_amount = min(num_builds, Configuration['pagination_max_limit'])
+
+        offset = offset if offset is not None else Configuration['pagination_offset']
+        limit = limit if limit is not None else Configuration['pagination_limit']
+
+        # Remove any negative values
+        offset = max(offset, 0)
+        limit = max(limit, 0)
+
+        # If limit is set higher than the number of builds/max limit, reduce limit
+        limit = highest_amount if limit > highest_amount else limit
+
+        # Requested offset too high should yield no results
+        if offset > num_builds:
+            return []
+
+        first_build_id = offset + 1
+        last_build_id = min((offset + limit), num_builds)
+
+        # Add +1 to last_build_id because range is exclusive
+        return [self.get_build(uid) for uid in range(first_build_id, last_build_id + 1)]
```

Comment:
Why use `get_build()` here instead of getting directly from `_all_builds_by_id`? It seems like code in this method is trying to make sure that every build in the computed range is a valid build id, so if there's a mistake in that logic it is more of a server error than a client error (so we don't need the potential ItemNotFoundError raised from `get_build()`.

Also, isn't this implementation a breaking change for v1 API's /builds endpoint? It seems like it would make that API stop returning the entire list of builds.